File: L07-StacksAndQueues/Brackets.java
// Date: Nov 28, 2021
// Brackets - Similar to https://leetcode.com/problems/valid-parentheses/
// Given S = "{[()()]}", the function should return 1
// Given S = "([)()]", the function should return 0

import java.util.*;

// time complexity: O(n)
// space complexity: O(n)
public int solution(String S) {
	Stack<Character> stk = new Stack<>();
	for (char ch: S.toCharArray()) {
		switch (ch) {
			case '{':
				stk.push('}');
				break;
			case '[':
				stk.push(']');
				break;
			case '(':
				stk.push(')');
				break;
			default:
				if (stk.empty() || ch != stk.pop())
					return 0;
		}
	}
	return stk.empty() ? 1 : 0;
}

// Reference: 
// https://stackoverflow.com/questions/196830/what-is-the-easiest-best-most-correct-way-to-iterate-through-the-characters-of-a

File: L07-StacksAndQueues/Brackets.swift
import Foundation
// result: https://app.codility.com/demo/results/trainingSQGNBM-4AC/
public class Brackets {
    public static func solution(_ S : inout String) -> Int {
        var stack = [Character]()
        for character in S {
            if character == "{" {
                stack.append("}")
            } else if character == "(" {
                stack.append(")")
            } else if character == "[" {
                stack.append("]")
            } else {
                if stack.isEmpty {
                    return 0
                }
                if let top = stack.popLast(), top != character {
                    return 0
                }
            }
        }
        return stack.isEmpty ? 1 : 0
    }
}

// reference
// https://stackoverflow.com/questions/25921204/convert-swift-string-to-array/25921323
